processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -97,19 +99,28 @@
       return String.class;
     }
 
+    /**
+     * numBytes will be -1 for null values.
+     */
     @Override
+    @Nullable
     public String fromByteBuffer(final ByteBuffer buffer, final int numBytes)
     {
-      return StringUtils.fromUtf8(buffer, numBytes);
+      // When SQL Compatibility is ON
+      //   1. numBytes will be -1 for null value, return value will be null
+      //   2. numBytes will be 0 for empty string, return value will be empty string
+      // For Legacy null handling when nulls and empty string are considered same -
+      //     numBytes will be 0 for both empty string and null and return value will be null for both.
```

Comment:
If so, could you please add something like `if (nullHandling.replaceWithDefaults()) throw new IllegalStateException("not expected")` in the `if (numBytes < 0) {` branch?

Reply:
Also, I think it's better to keep -1/null logic completely out of `ObjectStrategy` impls (like this one), and keep passing only non-negative numBytes into fromByteBuffer() method.

Reply:
done.